README.md

[![Arduino CI](https://github.com/RobTillaart/MTP40F/workflows/Arduino%20CI/badge.svg)](https://github.com/marketplace/actions/arduino_ci)
[![JSON check](https://github.com/RobTillaart/MTP40F/actions/workflows/jsoncheck.yml/badge.svg)](https://github.com/RobTillaart/MTP40F/actions/workflows/jsoncheck.yml)
[![Arduino-lint](https://github.com/RobTillaart/MTP40F/actions/workflows/arduino-lint.yml/badge.svg)](https://github.com/RobTillaart/MTP40F/actions/workflows/arduino-lint.yml)
[![License: MIT](https://img.shields.io/badge/license-MIT-green.svg)](https://github.com/RobTillaart/MTP40F/blob/master/LICENSE)
[![GitHub release](https://img.shields.io/github/release/RobTillaart/MTP40F.svg?maxAge=3600)](https://github.com/RobTillaart/MTP40F/releases)


# MTP40F

Arduino library for MTP40F CO2 sensor.


## Description

The library for the MTP40F CO2 sensor is **experimental** as not all functionality is tested.
It is based upon my https://github.com/RobTillaart/MTP40C library.

The MTP40F sensor is an NDIR (Non Dispersive InfraRed) CO2 sensor.

The sensor communicates over a 19200 baud serial (TTL) interface with a microprocessor or PC. 
This implies that calls which can take up to 25 bytes can take as much as about 20 milliseconds.

On the other hand this low baud rate implies it will work over relative long distances.
This signal quality over longer distances is not investigated. 

The MTP40F has more interface options, I2C, PWM and ALARM. 
This library does not support these other interfaces for now.
However minimal examples are added to have a starter but these 
need to be tested if and how well these work.


### Hardware interface MTP40-F  

Has TTL level RS232, I2C and PWM IO.

```
                        TOPVIEW MTP40-F
                       +-------------+
                       |             | 
    VCC (3v3 out)  5 --|             |-- 1 Vin
    TX             6 --|             |-- 2 GND
    RX             7 --|             |-- 3 ALARM
    NC             8 --|             |-- 4 PWM / I2C
    GND            9 --|             |
                       |             |
                       +-------------+
```

|  Pin  |  Name     |  Description                 |
|:-----:|:----------|:-----------------------------|
|   1   |  Vin      |  4.2V - 5.5V                 |
|   2   |  GND      |  idem                        |
|   3   |  ALARM    |  HIGH above 1000 PPM, LOW below 800 PPM (hysteresis) |
|   4   |  PWM/I2C  |  PWM out                     |
|   5   |  VCC_O    |  3V3 out for serial          |
|   6   |  TX       |  Transmit 9600 baud or SDA   |
|   7   |  RX       |  Receive 9600 baud or SCL    |
|   8   |  R/T      |  select I2C / Serial         |
|   9   |  GND      |  idem                        |


#### Links

- https://www.co2.earth/ - current outdoor CO2 level can be used for calibrating.
- https://keelingcurve.ucsd.edu/ - historical outdoor CO2 level.
- https://github.com/RobTillaart/MTP40C


## Interface

```cpp
#include "MTP40F.h"
```

#### Warnings

During tests with an UNO the communication over Software Serial did 
fail a few times.
Therefore it is important to **always check return values** 
to make your project more robust.

During tests it became clear that the sensor needs time to process 
commands e.g. **setSelfCalibration()**.

The CRC of the sensor responses are not verified by the library.


#### Constructors

- **MTP40F(Stream \* str)** constructor. should get a Serial port as parameter e.g. \&Serial, \&Serial1 
or a software Serial port. That Serial port must connect to the sensor. 
- **bool begin()** initialize the device.
Resets internal settings.
- **uint8_t getType()** returns type, see below.
Return 255 for the MTP40 base class.

|  Type  |  Model   |  Notes   |
|:------:|:--------:|:--------:|
|   2    |  MTP40C  |  for ref only
|   3    |  MTP40D  |  for ref only
|   5    |  MTP40F  |  
|  255   |  MTP40   |  for ref only


#### CO2 Measurement

- **uint16_t getGasConcentration()** returns the CO2 concentration in PPM (parts per million).
The function returns **MTP40_INVALID_GAS_LEVEL** if the request fails.
- **void suppressError(bool se)** sets or clears a flag that replaces the error value with 
the last read value if the request fails.
This is useful when plotting the values and one do not want a sudden spike.
One can still check **lastError()** to see if the value was OK.
- **bool getSuppressError()** gets the value of the suppress flag. 
- **int lastError()** returns last error set by **getGasConcentration()** 
or by **getAirPressureReference()** 
Reading resets internal error to MTP40F_OK;


#### Configuration

The library can set a maximum timeout in the communication with the sensor.
Normally this is not needed to set as the default of 100 milliseconds is long enough
for even the longest command. This timeout is needed if the sensor did not read the 
command correctly, preventing the host to wait indefinitely.
- **void setTimeout(uint32_t to = 100)** sets the timeout. 
If no parameter is given a default timeout of 100 milliseconds is set.
- **uint32_t getTimeout()** get the value set above or the default. 
Value returned is time in milliseconds.


## Calibration

Please read datasheet before using these functions to understand the process of calibration.

Note the outdoor calibration CO2 level differs per day and one should check 
a local airport or weather station for a good reference.

The University of San Diego keeps track of CO2 for a long time now.
See - https://keelingcurve.ucsd.edu/ 


#### Air pressure calibration

- **float getAirPressureReference()** returns the air pressure reference from the device.
Returns **MTP40F_INVALID_AIR_PRESSURE** in case request fails.
Default is 1013.0.
- **bool setAirPressureReference(float apr)** to calibrate the air pressure.
One can calibrate the sensor with an external device.
Value for air pressure should between 700.0 and 1100.0. 
The function returns **false** if the parameter is out of range or if the request fails.


#### SPC calibration

This takes a relative short time (few minutes) to calibrate the sensor in a known 
gas concentration. 

- **bool setSinglePointCorrection(uint32_t spc)** takes several minutes. see datasheet.
The parameter spc should be between 400 and 2000.
The function returns **false** if the parameter is out of range or if the request fails.
- **bool getSinglePointCorrectionReady()** To see if SPC is finished or not. 
The call also fails if the request fails.

As far as known the SPC point can not be retrieved from the sensor.


#### Self calibration

Self calibration is a process in which the sensor takes the minimum values over a longer period
between 24 - 720 hours as the reference for minimum outdoor values.
Note that 720 hours is 30 days / 1 month.

- **bool openSelfCalibration()** start the self calibration cycle.
- **bool closeSelfCalibration()** stop the self calibration cycle.
- **uint8_t getSelfCalibrationStatus()** Returns if the selfCalibration is open or closed.
**WARNING**: in our test the values in the datasheet seems to be not in sync with the sensor used. 
The function returned **0x00 for CLOSED and 0xFF for OPEN**.
- **bool setSelfCalibrationHours(uint16_t hrs)** Sets the number of hours between self calibration
moments. Valid values are 24 - 720 .
- **uint16_t getSelfCalibrationHours()** returns the value set above.


## Future

#### Must
- update documentation
  - reorder

#### Should 
- TODO in code
- ERROR handling.


#### Could 
- Performance
  - performance measurements
  - optimize performance if possible
  - caching? what?
  - seek optimum delay() between calls.
  - investigate wire length
- serial bus with multiple devices? => diodes
  - separate document
- improve readability code 
  - e.g. parameter names
- move all code from .h to .cpp file
- reuse command buffer as return buffer?
  - saves a bit.
- getAirPressureReference could be smarter 
  - always same value
  - from cache
  - dirty flag (-1);
- investigate timing of calls 
  - do they all need timeout of 4 seconds?

#### Wont

- store SPC point in the class?



CHANGELOG.md
# Change Log MTP40F

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/)
and this project adheres to [Semantic Versioning](http://semver.org/).


## [0.1.1] - 2023-08-02
- fix **getGasConcentration()** status handling
- update readme.md
- improve readability some parameters
- add (SoftwareSerial) schema to several examples
- add **MTP40F_REQUEST_FAILED** (error handling is still to do)
- add default value to **setAirPressureReference(apr = 1013)**
- add default value to **setSelfCalibrationHours(hours = 168)**


## [0.1.0] - 2023-07-25
- initial version.



test/unit_test_001.cpp
//
//    FILE: unit_test_001.cpp
//  AUTHOR: Rob Tillaart
// VERSION: 0.1.0
//    DATE: 2023-07-25
// PURPOSE: unit tests for the MTP40F CO2 sensor
//          https://github.com/RobTillaart/MTP40F
//          https://github.com/Arduino-CI/arduino_ci/blob/master/REFERENCE.md
//

// supported assertions
// https://github.com/Arduino-CI/arduino_ci/blob/master/cpp/unittest/Assertion.h#L33-L42
// ----------------------------
// assertEqual(expected, actual);               // a == b
// assertNotEqual(unwanted, actual);            // a != b
// assertComparativeEquivalent(expected, actual);    // abs(a - b) == 0 or (!(a > b) && !(a < b))
// assertComparativeNotEquivalent(unwanted, actual); // abs(a - b) > 0  or ((a > b) || (a < b))
// assertLess(upperBound, actual);              // a < b
// assertMore(lowerBound, actual);              // a > b
// assertLessOrEqual(upperBound, actual);       // a <= b
// assertMoreOrEqual(lowerBound, actual);       // a >= b
// assertTrue(actual);
// assertFalse(actual);
// assertNull(actual);

// // special cases for floats
// assertEqualFloat(expected, actual, epsilon);    // fabs(a - b) <= epsilon
// assertNotEqualFloat(unwanted, actual, epsilon); // fabs(a - b) >= epsilon
// assertInfinity(actual);                         // isinf(a)
// assertNotInfinity(actual);                      // !isinf(a)
// assertNAN(arg);                                 // isnan(a)
// assertNotNAN(arg);                              // !isnan(a)



////////////////////////////////////
//
//  MANY TESTS WILL BLOCK AS BUILD CI HAS NO GOOD TIMEOUT
//  ALL FAILING TESTS ARE COMMENTED
//
//  USE GODMODE SERIAL TO IMPROVE THESE TESTS
//  based upon MTP40 unit-tests

#include <ArduinoUnitTests.h>

#include "Arduino.h"
#include "MTP40F.h"
#include "SoftwareSerial.h"


unittest_setup()
{
  fprintf(stderr, "MTP40F_LIB_VERSION:\t%s\n", (char *) MTP40F_LIB_VERSION);
}


unittest_teardown()
{
}


unittest(test_constants)
{
  assertEqual(MTP40F_OK                  , 0x00);
  assertEqual(MTP40F_INVALID_AIR_PRESSURE, 0x01);
  assertEqual(MTP40F_INVALID_GAS_LEVEL   , 0x02);
  assertEqual(MTP40F_INVALID_ADDRESS     , 0xFF);
}


unittest(test_constructor)
{
  MTP40F sensor = MTP40F(&Serial);
  
  assertEqual(5, sensor.getType());
  assertTrue(sensor.begin());
  assertEqual(100, sensor.getTimeout());
  assertEqual(0, sensor.lastRead());
  assertEqual(MTP40F_OK, sensor.lastError());
}




unittest(test_air_pressure)
{
  MTP40F sensor = MTP40F(&Serial);

  assertEqual(MTP40F_INVALID_AIR_PRESSURE, sensor.getAirPressureReference());
  assertFalse(sensor.getSuppressError());
  sensor.suppressError(true);
  assertEqual(0, sensor.getAirPressureReference());
  sensor.suppressError(false);
  assertEqual(MTP40F_INVALID_AIR_PRESSURE, sensor.getAirPressureReference());

  assertFalse(sensor.setAirPressureReference(699.0));
  assertFalse(sensor.setAirPressureReference(1101));

  //  no communication
  assertFalse(sensor.setAirPressureReference(1000.0));
}


unittest(test_gas_concentration)
{
  MTP40F sensor = MTP40F(&Serial);

  assertEqual(MTP40F_INVALID_GAS_LEVEL, sensor.getGasConcentration());
  assertFalse(sensor.getSuppressError());
  sensor.suppressError(true);
  assertEqual(0, sensor.getGasConcentration());
  sensor.suppressError(false);
  assertEqual(MTP40F_INVALID_GAS_LEVEL, sensor.getGasConcentration());
}


unittest(test_single_point_correction)
{
  MTP40F sensor = MTP40F(&Serial);
  // assertTrue(sensor.begin());

  // assertFalse(sensor.getSinglePointCorrectionReady());

  assertFalse(sensor.setSinglePointCorrection(399));
  assertFalse(sensor.setSinglePointCorrection(2001));

  //  no communication
  assertFalse(sensor.setSinglePointCorrection(1000.0));
}


unittest(test_self_calibration)
{
  MTP40F sensor = MTP40F(&Serial);
  // assertTrue(sensor.begin());

  // assertFalse(sensor.openSelfCalibration());
  // assertFalse(sensor.closeSelfCalibration());
  // assertEqual(0x02, sensor.getSelfCalibrationStatus());
}


unittest(test_self_calibration_hours)
{
  MTP40F sensor = MTP40F(&Serial);
  // assertTrue(sensor.begin());

  assertFalse(sensor.setSelfCalibrationHours(23));
  assertFalse(sensor.setSelfCalibrationHours(721));
  // assertFalse(sensor.setSelfCalibrationHours(100));
  // assertEqual(0xFFFF, sensor.getSelfCalibrationHours());
}


unittest_main()


//  -- END OF FILE --



MTP40F.h
#pragma once
//
//    FILE: MTP40F.h
//  AUTHOR: Rob Tillaart
//    DATE: 2023-07-25
// VERSION: 0.1.0
// PURPOSE: Arduino library for MTP40F CO2 sensor
//     URL: https://github.com/RobTillaart/MTP40F


#include "Arduino.h"


#define MTP40F_LIB_VERSION                    (F("0.1.0"))


#define MTP40F_DEFAULT_ADDRESS                0x64

#define MTP40F_OK                             0x00
#define MTP40F_INVALID_AIR_PRESSURE           0x01
#define MTP40F_INVALID_GAS_LEVEL              0x02
#define MTP40F_INVALID_ADDRESS                0xFF


class MTP40F
{
public:
  MTP40F(Stream * str);

  bool     begin();

  int      getAirPressureReference();
  bool     setAirPressureReference(int apr);

  uint32_t getGasConcentration();  //  returns PPM

  void     suppressError(bool se) { _suppressError = se; };
  bool     getSuppressError()     { return _suppressError; };

  //  CALIBRATION FUNCTIONS
  //  READ DATASHEET !!
  bool     setSinglePointCorrection(uint32_t spc);
  bool     getSinglePointCorrectionReady();

  bool     openSelfCalibration();
  bool     closeSelfCalibration();
  uint8_t  getSelfCalibrationStatus();
  bool     setSelfCalibrationHours(uint16_t hrs);
  uint16_t getSelfCalibrationHours();


  //  set timeout of serial communication.
  void     setTimeout(uint32_t to = 100) { _timeout = to; };
  uint32_t getTimeout() { return _timeout; };

  uint32_t lastRead() { return _lastRead; };

  //       2 = MTP40C
  //       3 = MTP40D
  //       5 = MTP40F
  uint8_t  getType()  { return _type; };
  int      lastError();


//////////////////////////////////////////////////////


protected:
  Stream * _ser;
  uint8_t  _buffer[24];   //  should be big enough.

  uint32_t _timeout       = 100;
  uint32_t _lastRead      = 0;

  int      _airPressureReference = 0;
  uint32_t _gasLevel      = 0;
  uint8_t  _type          = 5;

  bool     _suppressError = false;
  int      _lastError     = MTP40F_OK;

  bool     request(uint8_t *data, uint8_t cmdlen, uint8_t anslen);
  uint16_t CRC(uint8_t *data, uint16_t len);
};


//  -- END OF FILE --



MTP40F.cpp
//
//    FILE: MTP40F.cpp
//  AUTHOR: Rob Tillaart
//    DATE: 2023-07-25
// VERSION: 0.1.0
// PURPOSE: Arduino library for MTP40F CO2 sensor
//     URL: https://github.com/RobTillaart/MTP40F


#include "MTP40F.h"

//  debug flag, development.
//  #define MTP40F_DEBUG    1


MTP40F::MTP40F(Stream * str)
{
  _ser = str;
  _buffer[0] = '\0';
  _type = 5;
}


bool MTP40F::begin()
{
  _timeout       = 100;
  _lastRead      = 0;
  _airPressureReference = 0;
  _gasLevel      = 0;
  _suppressError = false;
  _lastError     = MTP40F_OK;
  return true;
}


int MTP40F::getAirPressureReference()
{
  _lastError = MTP40F_OK;

  //  max read freq 1x per 4 seconds  //  TODO CHECK.
  if (millis() - _lastRead < 4000) return _airPressureReference;  //  last value
  _lastRead = millis();

  uint8_t cmd[9] = { 0x42, 0x4D, 0xA0, 0x00, 0x02, 0x00, 0x00, 0x01, 0x31 };
  if (request(cmd, 9, 11))
  {
    _airPressureReference = _buffer[7] * 256 + _buffer[8];
    return _airPressureReference;
  }

  _lastError = MTP40F_INVALID_AIR_PRESSURE;
  if (_suppressError) return _airPressureReference;
  return _lastError;
}


bool MTP40F::setAirPressureReference(int apr)
{
  if ((apr < 700) || (apr > 1100)) return false;  //  TODO CHECK

  uint8_t cmd[11] = { 0x42, 0x4D, 0xA0, 0x00, 0x01, 0x00, 0x02, 0x00, 0x00, 0x00, 0x00 };
  cmd[7] = apr / 256;
  cmd[8] = apr % 256;
  if (request(cmd, 11, 11) )
  {
    return true;
  }
  return false;
}


uint32_t MTP40F::getGasConcentration()
{
  _lastError = MTP40F_OK;

  //  max read freq 1x per 4 seconds
  if (millis() - _lastRead < 4000) return _gasLevel;  //  last value
  _lastRead = millis();

  uint8_t cmd[9] = { 0x42, 0x4D, 0xA0, 0x00, 0x03, 0x00, 0x00, 0x01, 0x32 };
  if (request(cmd, 9, 14) )
  {
    //  TODO  optimize shift.
    _gasLevel =  _buffer[7] * (256L * 256L * 256L);
    _gasLevel += _buffer[8] * (256L * 256L);
    _gasLevel += _buffer[9] * (256L);
    _gasLevel += _buffer[10];

    uint8_t status = _buffer[11];
    if (status == 0x00) return _gasLevel;
  }

  _lastError = MTP40F_INVALID_GAS_LEVEL;
  if (_suppressError) return _gasLevel;
  return _lastError;
}


bool MTP40F::setSinglePointCorrection(uint32_t spc)
{
  if ((spc < 400) || (spc > 2000)) return false;  //  TODO CHECK

  uint8_t cmd[13] = { 0x42, 0x4D, 0xA0, 0x00, 0x04, 0x00, 0x04, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00 };
  cmd[7]  = 0;
  cmd[8]  = 0;  
  cmd[9]  = spc / 256;
  cmd[10] = spc % 256;
  if (request(cmd, 13, 10) )
  {
    if (_buffer[7] ) return true;
  }
  return false;
}


bool MTP40F::getSinglePointCorrectionReady()
{
  uint8_t cmd[9] = { 0x42, 0x4D, 0xA0, 0x00, 0x05, 0x00, 0x00, 0x00, 0x00 };
  if (request(cmd, 9, 10) )
  {
    if (_buffer[8] == 0) return true;
  }
  return false;
}


bool MTP40F::openSelfCalibration()
{
  uint8_t cmd[10] = { 0x42, 0x4D, 0xA0, 0x00, 0x06, 0x00, 0x01, 0x00, 0x00, 0x00 };
  if (request(cmd, 10, 9) )
  {
    return true;
  }
  return false;
}


bool MTP40F::closeSelfCalibration()
{
  uint8_t cmd[10] = { 0x42, 0x4D, 0xA0, 0x00, 0x06, 0x00, 0x01, 0xFF, 0x00, 0x00 };
  if (request(cmd, 10, 9) )
  {
    return true;
  }
  return false;
}


uint8_t MTP40F::getSelfCalibrationStatus()
{
  uint8_t cmd[9] = { 0x42, 0x4D, 0xA0, 0x00, 0x07, 0x00, 0x00, 0x00, 0x00 };
  if (request(cmd, 9, 10) )
  {
    return _buffer[7];   //  0x00 or 0xFF
  }
  return 0x02;  //  TODO define error constant
}


uint16_t MTP40F::getSelfCalibrationHours()
{
  uint8_t cmd[9] = { 0x42, 0x4D, 0xA0, 0x00, 0x08, 0x00, 0x00, 0x00, 0x00 };
  if (request(cmd, 9, 11) )
  {
    uint16_t hrs = _buffer[7] * 256 + _buffer[8];
    return hrs;
  }
  return 0xFFFF;  //  TODO define error constant
}


bool MTP40F::setSelfCalibrationHours(uint16_t hrs)
{
  if ((hrs < 24) || (hrs > 720)) return false;
  uint8_t cmd[11] = { 0x42, 0x4D, 0xA0, 0x00, 0x09, 0x00, 0x02, 0x00, 0x00, 0x00, 0x00 };
  cmd[7] = hrs / 256;
  cmd[8] = hrs & 0xFF;
  if (request(cmd, 11, 10) )
  {
    return (_buffer[7] == 0x00);
  }
  return false;
}


int MTP40F::lastError()
{
  int e = _lastError;
  _lastError = MTP40F_OK;
  return e;
}


//////////////////////////////////////////////////////////////////////
//
//  PRIVATE
//
bool MTP40F::request(uint8_t *data, uint8_t commandLength, uint8_t answerLength)
{
  //  calculate CRC of command
  uint16_t crc = CRC(data, commandLength - 2);
  //  Serial.println(crc, HEX);
  data[commandLength - 2] = crc / 256;
  data[commandLength - 1] = crc & 0xFF;
  while (commandLength--)
  {
#ifdef MTP40F_DEBUG
    if (*data < 0x10) Serial.print(0);
    Serial.print(*data++, HEX);
    Serial.print(" ");
#else
    _ser->write(*data++);
#endif
     yield();   //  because baud rate is low!
  }

  //  Serial.print("XX: ");
  uint32_t start = millis();
  uint8_t i = 0;
  while (i < answerLength)
  {
    if (millis() - start > _timeout) return false;
    if (_ser->available())
    {
      _buffer[i] = _ser->read();
      //  Serial.print(_buffer[i], HEX);
      //  Serial.print(" ");
      i++;
    }
    yield();  //  because baud rate is low!
  }
  //  Serial.println();
  //  Serial.println(i);
  return true;
}


uint16_t MTP40F::CRC(uint8_t *data, uint16_t len)
{
  uint16_t sum = 0;
  for (uint16_t i= 0; i < len; i++)
  {
    sum += *data++;
  }
  return sum;
}


//  -- END OF FILE --

